### PR TITLE
Fix issue 20021 - static if doesn't evaluate opCast(bool)

### PR DESCRIPTION
**Pull request by dkorpel**

### CODE REVIEW

File: src/dmd/staticcond.d [hunk 1]
```diff
@@ -96,6 +96,8 @@ bool evalStaticCondition(Scope* sc, Expression original, Expression e, out bool
 
         e = e.expressionSemantic(sc);
         e = resolveProperties(sc, e);
+        if (e)
```
Comment (thewilsonator) [inline src/dmd/staticcond.d]: is this test necessary? its used unconditionally on [line 103](https://github.com/dlang/dmd/pull/10237/files#diff-d9f609df174a1d074cfc37b54558e504R103)

**Comment by thewilsonator** [comment]
Please update `ice11553.d` and target stable, this is fairly obviously not going to break anything.
```
Test fail_compilation/ice11553.d failed: 
expected:
----
fail_compilation/ice11553.d(22): Error: recursive template expansion while looking for `A!().A()`
----
actual:
----
fail_compilation/ice11553.d(22): Error: recursive template expansion while looking for `A!().A()`
fail_compilation/ice11553.d(22): Error: expression `A()` of type `void` does not have a boolean value
----
```

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/staticcond.d (hunk 1 above)
Comment (dkorpel) [reply src/dmd/staticcond.d]: I removed it. I'm a bit defensive while I'm still surprised by the abundant usage of null in dmd.

**Comment by dkorpel** [comment]
I tried targeting stable and rebasing with
```
git rebase --onto 062d2a origin/stable
git rebase --onto origin/stable master
```
But I'm still not familiar with rebasing, I don't think it worked out and now dlang-bot wants to auto-close 3 unrelated issues found in stable... How do you unlink issues?

**Comment by wilzbach** [comment]
> I tried targeting stable and rebasing with
> 
> ```
> git rebase --onto 062d2a origin/stable
> git rebase --onto origin/stable master
> ```
> 
> But I'm still not familiar with rebasing, I don't think it worked out and now dlang-bot wants to auto-close 3 unrelated issues found in stable... How do you unlink issues?

You don't. You need to do a rebase onto `upstream/stable`. That's all.

**Comment by dkorpel** [comment]
Ok I don't get git rebasing at all, this PR has become such a mess that I'm going to close it and revive it. Sorry for the inconvenience.

**Comment by ZombineDev** [comment]
@dkorpel The right way to go is the following:
* Ensure you gave the following 2 `git remote`s configured:
  * `upstream`, pointing to `git@github.com:dlang/dmd.git`
  * `origin`, pointing to your own fork of dmd, presumably `git@github.com:dkorpel/dmd.git`, from which you are going to open a pull request
  ```sh
  # Check which remotes you have configured:
  git remote -v

  # Add the following if missing:
  git remote add upstream git@github.com:dlang/dmd.git

  # If `git remote -v` had listed upstream, you need to
  # modify it, instead of adding it again:
  git remote set-url upstream git@github.com:dlang/dmd.git

  # (Do the same for origin.)

  # (If you don't have SSH Keys configured you may need to replace
  # those URLs with the https ones.)
  ```
* Before doing anything, ensure that your local `master` and `stable` branches are up to date with `upstream`:
```sh
git checkout master
git pull --ff-only upstream master

git checkout stable
git pull --ff-only upstream stable
```
* If you originally made your branch based of `$BRANCH_SOURCE`, and you need to rebase it on top of `$BRANCH_DESTINATION`, do the following:
```sh
# For your case the following may be the right branches:
BRANCH_SOURCE='master'
BRANCH_DESTINATION='upstream/stable`
PR_BRANCH='static-if-bool'

git checkout $PR_BRANCH
git rebase --onto $BRANCH_DESTINATION $BRANCH_SOURCE
```

**Pull request closed without merging**